packages/tasit-identity-contract/src/GnosisSafe.test.js

```diff
@@ -60,138 +61,211 @@ describe("GnosisSafe", () => {
+    it("contract-based account should send ERC20 tokens to someone", async () => {
+      const tokenAddress = ERC20_ADDRESS;
+      const { address: toAddress } = someone;
 
-      const { address: toAddress } = johnWallet;
-      const value = ONE;
+      gnosisSafe.setSigners([gnosisSafeOwner]);
+      gnosisSafe.setWallet(gnosisSafeOwner);
+      const action = gnosisSafe.transferERC20(tokenAddress, toAddress, ONE);
+      await action.waitForNonceToUpdate();
 
-      gnosisSafe.setWallet(johnWallet);
-      const execTxAction = gnosisSafe.transferEther(toAddress, value);
+      confirmBalances(erc20, [GNOSIS_SAFE_ADDRESS], [ZERO]);
+      confirmBalances(erc20, [toAddress], [ONE]);
+    });
 
-      const errorListener = async message => {
-        onError();
-        execTxAction.unsubscribe();
-      };
+    describe("spending by an ephemeral account", () => {
+      beforeEach("ethers to the ephemeral account pay for gas", async () => {
+        const { address: ephemeralAddress } = ephemeralAccount;
+        await etherFaucet(provider, root, ephemeralAddress, SMALL_AMOUNT);
+        confirmEtherBalances(provider, [ephemeralAddress], [SMALL_AMOUNT]);
+      });
+
+      it("ephemeral account shouldn't be able to transfer funds from contract-based account without allowance", async () => {
+        const onError = sinon.fake();
+        const balanceBefore = await erc20.balanceOf(GNOSIS_SAFE_ADDRESS);
+
+        const { address: toAddress } = someone;
+
+        gnosisSafe.setWallet(ephemeralAccount);
+        const action = erc20.transferFrom(GNOSIS_SAFE_ADDRESS, toAddress, ONE);
+
+        const errorListener = async message => {
+          onError();
+          action.unsubscribe();
+        };
+
+        // Note: Some error events are been trigger only from the confirmationListener
+        // See more: https://github.com/tasitlabs/TasitSDK/issues/253
+        const confirmationListener = () => {};
+
+        action.on("error", errorListener);
+        action.on("confirmation", confirmationListener);
+
+        await action.waitForNonceToUpdate();
+
+        await mineBlocks(provider, 1);
+
+        expect(onError.callCount).to.equal(1);
+        confirmBalances(erc20, [GNOSIS_SAFE_ADDRESS], [balanceBefore]);
+      });
+
+      describe("test cases that need ERC20 spending approval for ephemeral account", () => {
+        // TODO: Move to tasit-link-wallet
+        beforeEach(
+          "contract-based accounts' owner should approve an ephemeral account to spend funds",
+          async () => {
+            const tokenAddress = ERC20_ADDRESS;
+            const { address: spenderAddress } = ephemeralAccount;
+
+            gnosisSafe.setSigners([gnosisSafeOwner]);
+            gnosisSafe.setWallet(gnosisSafeOwner);
+            const action = gnosisSafe.approveERC20(
+              tokenAddress,
+              spenderAddress,
+              SMALL_AMOUNT
+            );
+            await action.waitForNonceToUpdate();
```

Comment:
So the action is still being sent upon building the action for now, right?

Reply:
Sorry, I didn't get it.

Reply:
I guess what I mean is, we changed the Gnosis-Safe-related actions to have a sync api, but we still haven't built them such that there's a second `.send` command that sends them - they still are automatically sent and then waiting for transaction confirmations is all that is necessary for them to complete, right? I'm 99% sure that's the case, just confirming since I wondered about it when reading the line above this one (that you can see part of on GitHub)

Reply:
Oh, got it now. Yes, that's the case.
This feature is tracked by https://github.com/tasitlabs/TasitSDK/issues/162 (I've just set the milestone since it hadn't one).

Reply:
Got it - thanks!

Reply:
Resolving this.